### PR TITLE
Refactor: extract worker process into separate class [changelog skip]

### PR DESCRIPTION
**Pull request by cjlarose**
### Description

My original motivation for extracting code related to the worker processes is the change described in https://github.com/puma/puma/issues/1875#issuecomment-695143880. I want to be able to create the first worker process (assuming `fork_worker` is on) by first `fork`ing from the master, then immediately `exec`ing into a new process definition with a separate memory space. 

My proof-of-concept in that comment technically _works_ but it's messy because the logic used by the worker process is woven in together with the logic used by the puma master process in `Puma::Cluster`. Worker processes implicitly rely on several instance variables used in `Puma::Cluster`, so for the new process to spin up a new worker, it needs to first spin up a dummy instance of `Puma::Cluster`.

To make it a little bit cleaner, I pulled out the worker logic into a new `Puma::Cluster::Worker` class. The initializer lays out everything that's necessary to construct a new worker process. My hope is that it's clear that this refactor is worth having regardless of the original motivation for it because it makes dependencies of the worker process clear and lowers the cognitive overhead of having worker logic and cluster logic in the same class definition.

### Your checklist for this pull request

- [x] I have reviewed the [guidelines for contributing](../blob/master/CONTRIBUTING.md) to this repository.
- [x] I have added an entry to [History.md](../blob/master/History.md) if this PR fixes a bug or adds a feature. If it doesn't need an entry to HISTORY.md, I have added `[changelog skip]` or `[ci skip]` to the pull request title.
- [x] I have added appropriate tests if this PR fixes a bug or adds a feature.
- [ ] My pull request is 100 lines added/removed or less so that it can be easily reviewed.
- [x] If this PR doesn't need tests (docs change), I added `[ci skip]` to the title of the PR.
- [x] If this closes any issues, I have added "Closes `#issue`" to the PR description or my commit messages.
- [x] I have updated the documentation accordingly.
- [x] All new and existing tests passed, including Rubocop.

### CODE REVIEW

**Comment by MSP-Greg** [comment]
@cjlarose 

Thank you for your work on this.

> cognitive overhead

Nice way of saying...  Having recently spent a few weeks away from thinking about anything Puma, I would agree.

Any thoughts on removing the current `Puma::Cluster::Worker` code from `lib/puma/cluster.rb` and moving it to `lib/puma/cluster/worker.rb` (same namespace of `Puma::Cluster::Worker`) and naming your file `lib/puma/cluster/worker0.rb`, with a namespace of `Puma::Cluster::Worker0`, or something like that?  Just a thought...

Being unable to run fork locally (hope to fix that soon), I can't really test this, but CI is passing.  👍

Lastly, the gem version/update issue would be helped by this.  If `autoload` was used for gems (both external and std-lib),  and maybe Puma files, it might remove issues with "where do we need require statements, and what about CI?", etc

**Comment by cjlarose** [comment]
> Any thoughts on removing the current Puma::Cluster::Worker code from lib/puma/cluster.rb and moving it to lib/puma/cluster/worker.rb (same namespace of Puma::Cluster::Worker) and naming your file lib/puma/cluster/worker0.rb, with a namespace of Puma::Cluster::Worker0, or something like that? Just a thought...

I agree that `Puma::Cluster` is the right namespace for both the new worker class and the existing worker class, but I wasn't crazy about the names `Puma::Cluster::Worker` and `Puma::Cluster::Worker0`, especially since the latter represents _all_ workers, not just the first one. I ended up renaming `Puma::Cluster::Worker` to `Puma::Cluster::WorkerHandle` and making the new class `Puma::Cluster::Worker`. What do you think?

**Comment by MSP-Greg** [comment]
I was mostly concerned with the namespace issue, and the embedded class in Cluster.  As to `WorkerHandle`, I'm not really sure, as it's kind of an different situation.  I can't think of another similar naming issue elsewhere...

JFYI, there are intermittent failures with non MRI jobs in CI, especially on macOS.

AFK.  Thanks, Greg

**Comment by gingerlime** [comment]
I'm a complete outsider to the Puma codebase, but just tried to understand the naming difference you discussed... 

So if I understand this correctly the `WorkerHandle` is the process handle for the worker, which you can then kill, check for liveness etc, and the `Worker` itself contains the actual logic. Right?

What about renaming `Worker` -> `WorkerLogic`? You can then either keep `WorkerHandle` as is? or maybe rename to `WorkerProcHandle` just to be more explicit? because "handle" can feel a little vague without context...

apologies if this doesn't make any sense and I'm butting in to a discussion that's just way over my head :) I know naming is hard, but I hope that my outsider's perspective can be helpful to see things with a fresh outlook.

**Comment by nateberkopec** [comment]
I'm kind of shocked that this clean of an extract was possible. I'm not terribly bothered about the current naming. Need to run through it in more detail but in concept I'm 100% behind this.

File: lib/puma/cluster.rb [hunk 1]
```diff
@@ -248,113 +173,23 @@ def wakeup!
...
-      title += " [#{@options[:tag]}]" if @options[:tag] && !@options[:tag].empty?
-      $0 = title
-
-      Signal.trap "SIGINT", "IGNORE"
-      Signal.trap "SIGCHLD", "DEFAULT"
-
-      fork_worker = @options[:fork_worker] && index == 0
-
       @workers = []
-      if !@options[:fork_worker] || fork_worker
```
Comment (nateberkopec) [inline lib/puma/cluster.rb:261]: I never realized this conditional was always true 😞

File: lib/puma/cluster.rb [hunk 2]
```diff
@@ -248,113 +173,23 @@ def wakeup!
...
 
-      server.run.join while restart_server.pop
-
-      # Invoke any worker shutdown hooks so they can prevent the worker
-      # exiting until any background operations are completed
-      @launcher.config.run_hooks :before_worker_shutdown, index, @launcher.events
-    ensure
-      @worker_write << "t#{Process.pid}\n" rescue nil
-      @worker_write.close
+      new_worker = Worker.new index: index,
```
Comment (nateberkopec) [inline lib/puma/cluster.rb:188]: Pretty delicious interface here 👍

**Pull request merged**